File: agents/python/agent.py
import json
import time
import uuid
import platform
import requests
import zstandard as zstd

class ApexAgent:
    def __init__(self, ingest_url, api_key):
        self.ingest_url = ingest_url
        self.api_key = api_key
        self.os = platform.system()
        self.arch = platform.machine()

    def capture_exception(self, e, stack_trace):
        report = {
            "error_id": str(uuid.uuid4()),
            "message": str(e),
            "stack_trace": stack_trace,
            "timestamp": int(time.time()),
            "context": {
                "os": self.os,
                "arch": self.arch,
                "total_memory": 16000000000, # Mocked for prototype
                "free_memory": 8000000000,
                "battery_level": 1.0,
                "is_charging": True,
                "network_type": "wifi"
            }
        }
        
        # In a real enterprise agent, we'd use the actual Protobuf library.
        # For this tactical prototype, we simulate the 'DNA' batch.
        batch = {"reports": [report]}
        
        # 1. JSON (Simulating Proto encoding for now, or just send JSON if modified)
        data = json.dumps(batch).encode('utf-8')
        
        # 2. Compress (Zstd)
        cctx = zstd.ZstdCompressor()
        compressed = cctx.compress(data)
        
        # 3. Sync
        headers = {
            "X-Apex-API-Key": self.api_key,
            "Content-Type": "application/octet-stream"
        }
        
        try:
            print(f"🚀 Apex_Python: Syncing forensic trace {report['error_id']}...")
            resp = requests.post(self.ingest_url, data=compressed, headers=headers)
            print(f"✅ Apex_Python: Server Response: {resp.status_code}")
        except Exception as err:
            print(f"❌ Apex_Python: Sync failed: {err}")

if __name__ == "__main__":
    # Test simulation
    agent = ApexAgent("http://localhost:8081/ingest", "apex-prod-key-12345")
    
    try:
        print("--- Simulating Python Fatal Error ---")
        # Simulate a bug
        x = {}
        print(x["missing_key"])
    except Exception as e:
        import traceback
        agent.capture_exception(e, traceback.format_exc())

File: agents/node/agent.js
const os = require('os');
const { v4: uuidv4 } = require('uuid');
const fzstd = require('fzstd');

class ApexAgent {
    constructor(ingestUrl, apiKey) {
        this.ingestUrl = ingestUrl;
        this.apiKey = apiKey;
        this.os = os.platform();
        this.arch = os.arch();
    }

    /**
     * Capture an error and sync it to the Apex Command Center.
     * @param {Error} error - The error object to capture.
     * @param {string} stackTrace - Optional custom stack trace.
     */
    async captureException(error, stackTrace) {
        const report = {
            error_id: uuidv4(),
            message: error.message,
            stack_trace: stackTrace || error.stack,
            timestamp: Math.floor(Date.now() / 1000),
            context: {
                os: this.os,
                arch: this.arch,
                total_memory: os.totalmem(),
                free_memory: os.freemem(),
                battery_level: 1.0, // Mocked for prototype
                is_charging: true,
                network_type: "wifi"
            }
        };

        const batch = { reports: [report] };
        
        try {
            // 1. JSON Encode
            const data = Buffer.from(jsonStringify(batch));
            
            // 2. Compress (Zstd)
            const compressed = fzstd.compress(data);
            
            console.log(`🚀 Apex_Node: Syncing forensic trace ${report.error_id}...`);

            // 3. Sync to Command Center
            const response = await fetch(this.ingestUrl, {
                method: 'POST',
                headers: {
                    'X-Apex-API-Key': this.apiKey,
                    'Content-Type': 'application/octet-stream'
                },
                body: compressed
            });

            console.log(`✅ Apex_Node: Server Response: ${response.status}`);
        } catch (err) {
            console.error(`❌ Apex_Node: Sync failed: ${err.message}`);
        }
    }
}

// Helper to handle BigInt or other non-JSON types if necessary
function jsonStringify(obj) {
    return JSON.stringify(obj, (key, value) =>
        typeof value === 'bigint' ? value.toString() : value
    );
}

module.exports = { ApexAgent };

// Simulation if run directly
if (require.main === module) {
    const agent = new ApexAgent("http://localhost:8081/ingest", "apex-prod-key-12345");

    console.log("--- Simulating Node.js Fatal Error ---");
    try {
        // Simulate a bug
        const obj = {};
        console.log(obj.missing.property); // Causes TypeError
    } catch (e) {
        agent.captureException(e);
    }
}
